File: app/src/main/java/es/bsalazar/txuntxungma/app/rates/RatesViewModel.java
package es.bsalazar.txuntxungma.app.rates;

import android.arch.lifecycle.MutableLiveData;

import java.util.List;

import es.bsalazar.txuntxungma.app.base.BaseViewModel;
import es.bsalazar.txuntxungma.data.DataSource;
import es.bsalazar.txuntxungma.data.remote.FirestoreSource;
import es.bsalazar.txuntxungma.domain.entities.Rate;
import es.bsalazar.txuntxungma.utils.ShowState;

class RatesViewModel extends BaseViewModel {

    private MutableLiveData<List<Rate>> ratesLiveData = new MutableLiveData<List<Rate>>() {
    };

    RatesViewModel(DataSource dataSource) {
        super(dataSource);
    }

    private boolean initialize;

    void getRates() {
        loadingProgress.setValue(ShowState.SHOW);
        initialize = true;
        dataSource.getRates(new FirestoreSource.OnCollectionChangedListener<Rate>() {
            @Override
            public void onCollectionChange(List<Rate> collection) {
                if (initialize) {
                    loadingProgress.setValue(ShowState.HIDE);
                    ratesLiveData.setValue(collection);
                    initialize = false;
                }
            }

            @Override
            public void onDocumentAdded(int index, Rate object) {
                if (!initialize) {

                }
            }

            @Override
            public void onDocumentChanged(int index, Rate object) {
                if (!initialize) {

                }
            }

            @Override
            public void onDocumentRemoved(int index, Rate object) {
                if (!initialize) {

                }
            }
        });
    }

    public MutableLiveData<List<Rate>> getRatesLiveData() {
        return ratesLiveData;
    }
}

File: app/src/main/java/es/bsalazar/txuntxungma/app/rates/RatesViewModel.kt
package es.bsalazar.txuntxungma.app.rates

import android.arch.lifecycle.MutableLiveData

import es.bsalazar.txuntxungma.app.base.BaseViewModel
import es.bsalazar.txuntxungma.app.base.lists.BaseListViewModel
import es.bsalazar.txuntxungma.data.DataSource
import es.bsalazar.txuntxungma.data.remote.FirebaseResponse
import es.bsalazar.txuntxungma.data.remote.FirestoreSource
import es.bsalazar.txuntxungma.domain.entities.*
import es.bsalazar.txuntxungma.utils.ShowState

class RatesViewModel(dataSource: DataSource) : BaseListViewModel<Rate>(dataSource) {

    val authLiveData = object : MutableLiveData<Auth>() {}

    fun getRates() { execute(dataSource::getRates) }

    fun getLoginData() {
        dataSource.getLoginData(object : DataCallback<Auth, BaseError> {
            override fun onSuccess(response: Auth?) {
                authLiveData.value = response
            }

            override fun onFailure(errorBase: BaseError?) {
                authLiveData.value = null
            }
        })
    }

    fun saveRate(rate: Rate) = dataSource.saveRate(rate) {}

    fun modifyRate(rate: Rate) = dataSource.updateRate(rate) {}

    fun deleteRate(rate: Rate) = dataSource.deleteRate(rate.id)
}